### PR TITLE
feature/152638438 registration flow

### PR DESCRIPTION
**Pull request by epinxteren**

### CODE REVIEW

File: src/Features/Fixture/AuthnRequest_registration.xml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<samlp:AuthnRequest
+        AssertionConsumerServiceURL="https://pieter.aai.surfnet.nl/simplesamlphp/module.php/saml/sp/saml2-acs.php/default-sp"
+        Destination="https://gssp.stepup.example.com/app_dev.php/saml/sso"
+        ID="_a371d8461d33016b12dc843b41b65a1af997a41900" IssueInstant="2017-11-16T12:45:39Z"
+        ProtocolBinding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST" Version="2.0"
+        xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion"
+        xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol">
+    <saml:Issuer>https://pieter.aai.surfnet.nl/simplesamlphp/module.php/saml/sp/metadata.php/default-sp</saml:Issuer>
+    <samlp:NameIDPolicy AllowCreate="true" Format="urn:oasis:names:tc:SAML:2.0:nameid-format:transient"/>
+</samlp:AuthnRequest>
```
Comment (epinxteren) [inline src/Features/Fixture/AuthnRequest_registration.xml]: Can be deleted

File: src/Controller/IdentityController.php [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+        return $response;
+    }
+
+    /**
+     * @param string $view
+     * @param SAML2_Response $response
+     * @return Response
+     */
+    public function renderSamlResponse($view, SAML2_Response $response)
```
Comment (MKodde) [inline src/Controller/IdentityController.php]: No need to make this method publicly available, right?

File: src/Service/ConfigurationContainer.php [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    private $registrationRoute;
+
+    /**
+     * @param array[] $configuration
+     *
+     * @throws \Assert\AssertionFailedException
+     */
+    public function __construct($configuration)
+    {
+//        Assertion::keyExists($configuration, 'registration_route');
```
Comment (MKodde) [inline src/Service/ConfigurationContainer.php]: Forgot to remove these comments?

File: src/Saml/ResponseBuilder.php [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        ], true);
+    }
+
+    private function isValidResponseSubStatus($subStatus)
+    {
+        return in_array($subStatus, [
+            SAML2_Const::STATUS_AUTHN_FAILED,               // failed authentication
+            SAML2_Const::STATUS_INVALID_ATTR,
+            SAML2_Const::STATUS_INVALID_NAMEID_POLICY,
+            SAML2_Const::STATUS_NO_AUTHN_CONTEXT,           // insufficient Loa or Loa cannot be met
```
Comment (MKodde) [inline src/Saml/ResponseBuilder.php]: Do you need to test for a valid Authn Context? In the case of Gateway the required Loa is loaded from the Authn Context. In this context this seems irrelevant. Please verify this with an expert ;)

File: README.md [hunk 5]
```diff
@@ -49,7 +49,14 @@ surfnet_saml:
             certificate_file: "%saml_remote_idp_certificate%" 
 ```
 
-See [Saml bundle documentation](https://github.com/OpenConext/Stepup-saml-bundle) for more information about this configuration
+See [Saml bundle documentation](https://github.com/OpenConext/Stepup-saml-bundle) for more information about the configuration above
+
+```yaml
+assetic:
```
Comment (MKodde) [inline README.md:55]: I thought you would have chosen the more modern [Encore](https://symfony.com/blog/introducing-webpack-encore-for-asset-management). (webpack based) frontend asset management solution. Im totaly fine with the Assetic solution, but you might want to reconsider.

File: README.md [hunk 6]
```diff
@@ -76,6 +83,8 @@ This is example idp configuration that works with [pieter.aai.surfnet.nl](https:
 Development environment
 ======================
 
+The purpose of the development environment is only for running the different test and metric tools only.
```
Comment (MKodde) [inline README.md]: Suggestion to remove the last 'only' from this sentence.

File: src/Controller/IdentityController.php [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+            ->setRequestServiceProvider($originalRequest->getServiceProvider())
+            ->setRelayState($request->get(AuthnRequest::PARAMETER_RELAY_STATE, ''));
+        return new RedirectResponse($this->generateUrl($this->registrationRoute->getRegistrationRoute()));
+    }
+
+    /**
+     * @Route("/saml/sso_return", name="gssp_saml_sso_return", methods={"POST", "GET"})
+     * @throws \Exception
+     */
+    public function ssoReturnAction()
```
Comment (MKodde) [inline src/Controller/IdentityController.php]: Consider describing the intent of this controller action. This comment goes for the other actions as well.

File: tests/TestUnitTest.php [hunk 8]
```diff
@@ -15,7 +15,7 @@
  * limitations under the License.
  */
 
-namespace Surfnet\GsspBundle;
+namespace Tests\Surfnet\GsspBundle;
```
Comment (MKodde) [inline tests/TestUnitTest.php]: This test can be removed as discussed IRL.

File: src/Features/Context/GsspContext.php [hunk 9]
```diff
@@ -17,14 +17,478 @@
...
+     *
+     * @throws \Exception
+     */
+    public function createANormalAuthnRequest()
+    {
+        $request = new SAML2_AuthnRequest();
+        $request->setAssertionConsumerServiceURL($this->serviceProvider->getAssertionConsumerUrl());
+        $request->setDestination($this->identityProvider->getSsoUrl());
+        $request->setIssuer($this->serviceProvider->getEntityId());
+        $request->setProtocolBinding(SAML2_Const::BINDING_HTTP_POST);
```
Comment (MKodde) [inline src/Features/Context/GsspContext.php]: Using the `SAML2_Const::BINDING_HTTP_REDIRECT` protocol binding is more in line with how this service will be used.

**Review by MKodde — CHANGES_REQUESTED** [review]

**Pull request merged**